### PR TITLE
Add support on iOS for snapshot.key, update and once.

### PR DESCRIPTION
**Pull request by pietgk**
The Android support is still WIP.

database.js:
```
  DataSnapshot.key()
  Query.once(eventType:EventType, cb:((snapshot:DataSnapshot) => Promise)) : () => void
  DatabaseReference.update(value:any) : Promise

```
iOS implementation in FirebaseBridgeDatabase.swift
```
  @objc func snapshotKey(snapshotUUID: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock)
  @objc func onRef(databaseUrl: String?, eventTypeString:String, query: [[AnyObject]]) throws -> FIRDatabaseQuery
  @objc func update(databaseUrl:String, value:Dictionary<String, AnyObject>, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock)

```
and the needed updates in FirebaseBridge.m

As simple test context an example of the code i used to get things going is included below:

```
import Database from 'rn-firebase-bridge/database';
import { signInWithEmail } from 'rn-firebase-bridge/auth';

export default class FirebaseNativeService {

  constructor(layerContext) {
    this.layerContext = layerContext;

    if (FirebaseNativeService.databasePersistenceEnabled === undefined) {
      FirebaseNativeService.databasePersistenceEnabled = true;

      Database.setPersistenceEnabled(true);

      this.signIn('xxx@yyy.nl', 'ppp')
        .then(() => {
          this.test();
        });
    }
  }

  async test() {
    console.log('========== test called');
    return this.testSetUpdateRemove()
      .then(() => console.log('========== testSetUpdateRemove ok'))
      .catch(err => console.log('!!!!!!!!!! testSetUpdateRemove error:', err));
  }

  async testSetUpdateRemove() {
    console.log('====----------====== testSetUpdateRemove started');

    const testListenDisposers = await this.testListen();

    await this.testSetValues();
    const testSetValuesOkDisposers = await this.testSetValuesOk();

    await this.testUpdateValues();
    const testUpdateValuesOkDisposers = await this.testUpdatedValuesOk();

    await this.testRemoveValues();
    await this.testRemovedValuesOk();

    this.stopListeningOn(testListenDisposers);
    this.stopListeningOn(testSetValuesOkDisposers);
    this.stopListeningOn(testUpdateValuesOkDisposers);

    this.newId('testFirebaseNativeService/newId');

    console.log('====----------====== testSetUpdateRemove finished');
  }

  testRefAaa() {
    const ref = this.dbRef();
    return ref
      .child('testFirebaseNativeService/Aaa');
  }

  async testListen() {
    console.log('========== testListen started');
    const ref = this.testRefAaa();
    ref.toString().then(result => console.log('testListen on:', result), err => console.log('testListen error:', err));

    const disposers = {};
    disposers.testListen = ref.on('value', async (snapshot) => {
      await snapshot.forEach(async (child) => {
        const val = await child.val();
        const key = await child.key();
        console.log('testListen Child key:', key, 'val:', val);
      });
      console.log('testListen Snapshot val:', await snapshot.val(), ', key:', await snapshot.key());
    });
    return disposers;
  }

  async testSetValues() {
    console.log('========== testSetValues started');
    const ref = this.testRefAaa();
    return ref
      .setValue({ val1: 'val1', val2: { val2a: 'val2a', val2b: 'val2b' } })
      .then(
        () => {
          console.log('testSetValues ok');
        },
        err => console.log('testSetValues err:', err));
  }

  async testSetValuesOk() {
    console.log('========== testSetValuesOk started');
    const ref = this.testRefAaa();
    ref.toString().then(result => console.log('testSetValuesOk testRefAaa toString:', result), err => console.log('testSetValuesOk toString error:', err));

    const disposers = {};
    disposers.testSetValuesOkOne = ref.once('value', async (snapshot) => {
      await snapshot.forEach(async (child) => {
        const val = await child.val();
        const key = await child.key();
        console.log('testSetValuesOk Child key:', key, 'val:', val);
      });
      console.log('testSetValuesOk Snapshot val:', await snapshot.val(), ', key:', await snapshot.key());
    });
    return disposers;
  }

  async testUpdateValues() {
    console.log('========== testUpdateValues started');
    const ref = this.testRefAaa();
    return ref
      .update({ val3: 'val3', val2: { val2a: 'val2aUpdated', val2b: 'val2bUpdated' } })
      .then(
        () => {
          console.log('testUpdateValues ok');
        },
        err => console.log('testUpdateValues err:', err));
  }

  async testUpdatedValuesOk() {
    console.log('========== testUpdateValuesOk started');
    const ref = this.testRefAaa();

    const disposers = {};
    disposers.testSetValuesOkOne = ref.once('value', async (snapshot) => {
      await snapshot.forEach(async (child) => {
        const val = await child.val();
        const key = await child.key();
        console.log('testUpdatedValuesOk Child key:', key, 'val:', val);
      });
      console.log('testUpdatedValuesOk Snapshot val:', await snapshot.val(), ', key:', await snapshot.key());
    });
    return disposers;
  }

  async testRemoveValues() {
    console.log('========== testRemoveValues started');
    const ref = this.testRefAaa();
    return ref.remove()
      .then(
        () => {
          console.log('testRemoveValues ok');
        },
        err => console.log('testRemoveValues err:', err));
  }

  async testRemovedValuesOk() {
    console.log('========== testRemovedValues started');
    const ref = this.testRefAaa();
    return ref.once('value', async (snapshot) => {
      console.log('testRemovedValuesOk Snapshot exists:', await snapshot.exists());
    });
  }

  signIn(email, password) {
    return signInWithEmail(email, password).then(
        user => {
          console.log(user);
        },
        error => console.log(error)
    );
  }

  async newId(modelPath) {
    const idRef = this.dbRef().child(modelPath);
    const entityIdRef = idRef.push();
    const key = await entityIdRef.key();
    console.log('newId:', key);
    return key;
  }

  dbRef() {
    if (this.rootRef === undefined) {
      this.rootRef = Database
        .ref()
        .child(this.layerContext.databaseRoot || 'qqqq');
    }
    return this.rootRef;
  }
}
```

### CODE REVIEW

File: database.js [hunk 1]
```diff
@@ -255,6 +303,13 @@ export class DatabaseReference extends Query {
             ({ locationUrl }) => NativeFirebaseBridgeDatabase.setValue(locationUrl, [value]));
     }
 
+    update(value:any) : Promise {
+        // We wrap value in array for easier handling on Android.
+        // See FirebridgeDatabase.java setValue()
```
Comment (davecoates) [inline database.js]: this comment isn't relevant

File: database.js [hunk 2]
```diff
@@ -255,6 +303,13 @@ export class DatabaseReference extends Query {
             ({ locationUrl }) => NativeFirebaseBridgeDatabase.setValue(locationUrl, [value]));
     }
 
+    update(value:any) : Promise {
```
Comment (davecoates) [inline database.js]: type should be `value:{ [key: string]: any }`

File: ios/FirebaseBridgeDatabase.swift [hunk 3]
```diff
@@ -191,9 +241,20 @@ class FirebaseBridgeDatabase: NSObject, RCTInvalidating {
     resolve(convertRef(getRefFromUrl(databaseUrl).childByAutoId()))
   }
   
+  // We receive an array of a single element whh is the value to set
```
Comment (davecoates) [inline ios/FirebaseBridgeDatabase.swift]: comment not relevant to this function

**Comment by davecoates** [comment]
Thanks very much for this!

Are you working on the Android version, or should I add that?

**Comment by pietgk** [comment]
Hello Dave,

if you could add the android version then that would be very much appreciated.

i am working on getting FireBaseStorage added.

This to get everything working for iOS first.

Then later ill start to get it working on android.

What is the timezone you live in? mine is amsterdam gmt+1

**Comment by davecoates** [comment]
Hi Piet,

No problem, I'll look at doing it later this week.

Ok great, that sounds good

I'm in Melbourne, GMT +10

**Comment by davecoates** [comment]
Android support added

**Comment by pietgk** [comment]
Cool thanks very much !!

i am testing and using a Storage API with only a simple fetch into a local file (all i need right now)

**Pull request merged**